### ISSUE TITLE
PageIndex Inaccessible Dropdown Menu Pattern

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#783]
For each problem set in the list, a toggle control is present allowing a sub-menu to be expanded.  The sub-menu contains the “Duplicate Set” and “Share with Teachers” controls.  Both the toggle button and sub-menu content are completely inaccessible to, and unusable by, all keyboard users, including those operating a screen reader. 
 
Firstly, the toggle < button> is a child of the problem set link, as is the < div> containing the sub-menu items.  Those items are themselves marked up as links, meaning that there is one < button> and two < a> elements marked up as children of a < button>.  The HTML Specification (https://html.spec.whatwg.org/multipage/) defines a content model (https://html.spec.whatwg.org/multipage/dom.html#concept-element-content-model) for all elements, described as: “A normative description of what content must be included as children and descendants of the element”.  For both < a> (defined in 4.5.1) and < button> (defined in 4.10.6), the content model clearly states that “there must be no interactive content descendant”. 
 
In short, this sub-menu implementation pattern does not represent valid HTML.  A browser will not expose the child <button> toggle control in the accessibility tree, so a screen reader user won’t even know that the sub-menu exists.  Meanwhile, people with some usable vision who still rely on the keyboard (such as users of screen magnification) may be able to see the toggle control, but become bewildered when they can’t set focus to or activate it. 
 
Next, the mark-up of the toggle button and sub-menu content presents some accessibility concerns: 
 
• The toggle button has no accessible label.  Visually, it is labelled using a Font Awesome icon, but no textual alternative is implemented. 
• An aria-haspopup attribute (with a value of “true”) has been applied to the toggle button.  This is incorrect because, while the button does control the display of sub-menu-style content, the sub-menu itself is not (and should not) be marked up as a menu proper. 
• The sub-menu is contained inside a < div> instead of a < ul>. 
• The sub-menu’s container has an aria-labelledby attribute applied to provide an accessible name, but its value references a non-existent element ID. 
 
Finally, users who are able to operate the menu using a mouse may become confused, when they find that activating the “Share with Teachers” option actually opens the Twitter website with a pre-filled tweet.

**Comment by johnhbenetech** [comment]
Suggested Solution 
 
Nesting interactive controls inside other interactive controls in this way must be avoided.  Here, the button to toggle the sub-menu, and the sub-menu itself, should be repositioned in the DOM to follow the problem set link instead of being children of it.  In addition, it would be appropriate to avoid using a < button> element for the problem set link, and then overriding its accessible role.  The control should be  a hyperlink, so there’s no reason not to use a < a> directly. 
 
An accessible label must then be applied to the toggle, which can be accomplished using an aria-label attribute.  The text can be something like: “More Options”, and should include the name of the corresponding problem set to avoid multiple buttons with exactly the same label but different contexts.  For example: “More options for Distance, Rate and Time”.  Of note is that the number of problems in the set, which is present in the problem set link text, should not be included in the toggle button label.  This string will also need to be localized. 
 
Thirdly, unnecessary and problematic attributes should be removed from all < button> and < a> controls.  This includes the aria-haspopup attribute from the toggle button, as well as the tabindex and type attributes from all other < button> and < a> elements. 
 
Next, the sub-menu should be marked up as an unordered list instead of a flat < div>, with the two child links wrapped inside < li> elements.  The value of the aria-labelledby attribute on the opening < ul> should be corrected to refer to the ID of the corresponding toggle button. 
 
And finally, it would be helpful to modify the link text of the “Share with Teachers” option so that it more accurately reflects its true purpose and destination.  For example, the text could just read “Share on Twitter”.  This change will require  updating the share_with_teachers key in the strings.js file.

**Comment by johnhbenetech** [comment]
Proposed Code Sample 
 
File: src/components/PageIndex/index.js 

```
<li className="card" key={index}>  <a   href="#"   className="btn d-flex"   onClick={this.openPremadeSet(exampleProblemSet)}   onKeyPress={this.openPremadeSet(exampleProblemSet)}  >   <span className={pageIndex.title}>{exampleProblemSet.title}</span>   <span className={pageIndex.meta}>&nbsp;({exampleProblemSet.problems.length} Problems)</span>  </a>  <div className="dropdown">   <button    className={`btn dropdown-toggle ${pageIndex.problemSetDropdown}`}    id={`dropdownMenuButton-${index}`}    data-toggle="dropdown"    aria-expanded="false"    aria-label={`More options for {exampleProblemSet.title}`}    onClick={(e) => {     stopEvent(e);    }}   >    <FontAwesome name="ellipsis-v" />   </button>   <ul className="dropdown-menu" arialabelledby={`dropdownMenuButton-${index}`}>    <li>     <a      href="#"      className="dropdown-item"      onClick={this.duplicateProblemSet(exampleProblemSet)}      onKeyPress={this.duplicateProblemSet(exampleProblemSet)}     > 
{` ${Locales.strings.duplicate_set}`}      <FontAwesome size="lg" name="copy" />     </a>     <a      href="#"      className="dropdown-item"      onClick={shareOnTwitter(exampleProblemSet.shareCode)}      onKeyPress={shareOnTwitter(exampleProblemSet.shareCode)}     >      {` ${Locales.strings.share_with_teachers}`}      <FontAwesome size="lg" name="twitter" />      </a>    </li>   </ul>  </div> </li>
```

**Comment by rupeshparab** [comment]
This has been fixed in #752

**Comment by jscholes** [comment]
@rupeshparab - the dropdown menu and triggering button are still wrapped inside the link for the problem set.  The `<button>` and dropdown should be contained inside the same `<li>`, but they cannot be inside the `<a>`.  As indicated on #752, the `aria-haspopup` attribute should also be removed from the dropdown trigger.

**Comment by rupeshparab** [comment]
I have moved dropdown out of the anchor

**Comment by jscholes** [comment]
@rupeshparab - great.  All we need to do now is give each button a unique accessible label ("More Options for [set title]"), and fix the `aria-labelledby` on each `<ul>` to reference the ID of the dropdown button.

**Comment by rupeshparab** [comment]
I have updated the label to include the title and also fixed the aria-label ids

**Comment by jscholes** [comment]
@rupeshparab - this seems to be working as intended, in all but one case.  The button for the first problem set in the list correctly has a label of "More Options for Distance Rate and Time",, but the list on expansion has the label "More Options for Example Problem Set".

**Comment by jscholes** [comment]
Incorrectly closed.

**Comment by rupeshparab** [comment]
@jscholes 
I have pushed a fix for this

**Comment by jscholes** [comment]
@rupeshparab - great, thanks!  Could we remove the code which is causing menus to be expanded based on a user's focus movements?  There should only be two ways to collapse one of these menus: explicitly press the button again, or press another button to expand a different one.  Focus movements alone (including activating the share controls) shouldn't cause them to be unexpectedly collapsed.

**Comment by rupeshparab** [comment]
When I move focus on these buttons, they are not toggling the dropdown menu until I press Space or Enter.
Can you share the browser which you are using and its version?

**Comment by jscholes** [comment]
@rupeshparab - Chrome, version 78.0.3904.87 (Official Build) (64-bit).

To reproduce:

1. Open the "All Problem Sets" page.
2. Expand the options menu for "Distance Rate and Time".
3. With focus on the options button, press Shift+Tab.  Focus moves to the problem set link, and the menu collapses.
4. Expand the menu again.
5. With focus on the options button, press Tab three times.  Focus moves to the "Combining Like Terms" set link, and the menu for "Distance Rate and Time" collapses.
6. Expand the menu again.
7. Tab to either the "Duplicate Set" or "Share on Twitter" options and activate it.
8. Close the resulting tab.  The menu has been collapsed.

**Comment by rupeshparab** [comment]
Got it now, currently there isn't any code which is manually closing the dropdown.
I am investigating whether something needs to be configured in bootstrap to prevent dropdown from closing.

**Comment by rupeshparab** [comment]
The dropdowns won't close on losing focus now

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/ef5c95051cd982926639e36662ab93f8f8605de3.
